src/mlpack/methods/ann/layer/adaptive_max_pooling.hpp

```diff
@@ -0,0 +1,172 @@
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class AdaptiveMaxPooling
+{
+ public:
+  //! Create the AdaptiveMaxPooling object.
+  AdaptiveMaxPooling();
+
+  /**
+   * Create the AdaptiveMaxPooling object.
+   *
+   * @param outputWidth Width of the output.
+   * @param outputHeight Height of the output.
+   */
+  AdaptiveMaxPooling(const size_t outputWidth,
+                     const size_t outputHeight);
+
+  /**
+   * Create the AdaptiveMaxPooling object.
+   *
+   * @param outputShape A two-value tuple indicating width and height of the output.
+   */
+  AdaptiveMaxPooling(const std::tuple<size_t, size_t> outputShape);
+
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename eT>
+  void Forward(const arma::Mat<eT>& input, arma::Mat<eT>& output);
+
+  /**
+   * Ordinary feed backward pass of a neural network, using 3rd-order tensors as
+   * input, calculating the function f(x) by propagating x backwards through f.
+   * Using the results from the feed forward pass.
+   *
+   * @param input The propagated input activation.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>& input,
+                arma::Mat<eT>& gy,
+                arma::Mat<eT>& g);
+
+  //! Get the output parameter.
+  const OutputDataType& OutputParameter() const
+  { return poolingLayer.OutputParameter(); }
+
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return poolingLayer.OutputParameter(); }
+
+  //! Get the delta.
+  const OutputDataType& Delta() const { return poolingLayer.Delta(); }
+  //! Modify the delta.
+  OutputDataType& Delta() { return poolingLayer.Delta(); }
+
+  //! Get the width.
+  size_t InputWidth() const { return poolingLayer.InputWidth(); }
+  //! Modify the width.
+  size_t& InputWidth() { return poolingLayer.InputWidth(); }
+
+  //! Get the height.
+  size_t InputHeight() const { return poolingLayer.InputHeight(); }
+  //! Modify the height.
+  size_t& InputHeight() { return poolingLayer.InputHeight(); }
+
+  //! Get the width.
+  size_t OutputWidth() const { return outputWidth; }
+  //! Modify the width.
+  size_t& OutputWidth() { return outputWidth; }
+
+  //! Get the height.
+  size_t OutputHeight() const { return outputHeight; }
+  //! Modify the height.
+  size_t& OutputHeight() { return outputHeight; }
+
+  //! Get the reset parameter of the layer.
+  bool Reset() const { return reset; }
+  //! Modify the reset parameter of the layer.
+  bool& Reset() { return reset; }
```

Comment:
reset should be private no need for getters and setters

Reply:
Ohk got it. But then one can't modify the layer for different size but  I guess that's okay.

Reply:
hmm. IMO `reset` is meant only for internal use in the layer and therefore it should not be exposed as a part of the API.
> But then one can't modify the layer for different size but I guess that's okay.

Yes I was also considering that. But if the layer is used as a part of `FFN` then the ouputWidth and outputHeight of this layer should not change and also the inputWidth and inputHeight to the layer will remain fixed.

Reply:
Agreed, that makes sense. I have made the changes. Thanks.